doc/design/listener.md

```diff
@@ -0,0 +1,323 @@
+|   - m_events : Event_t[]                        | |   void getUniqueId()           |
+|   - m_thread : std::thread                      | |                                |
+|   - m_conditionListener : ConditionListener     | |   - m_conditionVariableDataPtr |
+|                                                 | |   - m_resetCallback            |
+| +----------------------------+                  | |   - m_uniqueTriggerId          |
+| | Event_t                    |                  | +--------------------------------+
+| |   void executeCallback()   |                  |      | 1
+| |   bool reset()             |                  |      |
+| |   bool init(...)           |                  |      | n
+| |                            |                  | +-------------------------------------------------------+
+| |   - m_origin               |                  | | Triggerable (e.g. Subscriber)                         |
+| |   - m_callback             |                  | |                                                       |
+| |   - m_invalidationCallback |                  | |   void invalidateTrigger(const uint64_t triggerId)    |
+| |   - m_eventId              |                  | |   void enableEvent(TriggerHandle&&, const EventEnum ) |
+| +----------------------------+                  | |   void enableEvent(TriggerHandle&&)                   |
++-------------------------------------------------+ |   void enableState(TriggerHandle&&, const StateEnum)  |
+                                                    |   void enableState(TriggerHandle&&)                   |
+                                                    |                                                       |
+                                                    |   void disableEvent(const EventEnum)                  |
+                                                    |   void disableEvent()                                 |
+                                                    |   void disableState(const StateEnum)                  |
+                                                    |   void disableState()                                 |
+                                                    |                                                       |
+                                                    |   - m_triggerHandle : TriggerHandle                   |
+                                                    +-------------------------------------------------------+
+```
+
+The Triggerable does not need to implement all `enable{Event,State}`,
+`disable{Event,State}` variations only the ones which are required by the use
+case. If the Triggerable for instance has no triggering states then
+`enableState`/`disableState` can be omitted. When there is only a single state
+or event which can be triggered then the `enable{Event,State}` and
+`disable{Event,State}` without the distincting `{State,Event}Enum` can be used.
+
```

Comment:
Not sure if we should mention the `enable/disableState` methods here since this is the design document for the event-driven listener.